src/funcade/jwks.clj

```diff
@@ -2,39 +2,77 @@
   (:require [jsonista.core :as json]
             [buddy.core.keys :as bk]
             [org.httpkit.client :as http]
-            [jsonista.core :as j]
-            [funcade.tools :as t]))
+            [funcade.tools :as t]
+            [yang.scheduler :as scheduler]))
 
 (defonce ^:private
-  keyset (atom {}))
+  keyset (atom nil))
+
+(defn- find-token-key-by-kid
+  "Given the token's kid find the key from keyset"
+  [& kids]
+  (when (and (some? @keyset)
+             (seq kids))
+    (get-in @keyset kids)))
+
+(defn find-current-kids
+  "Find all current kids"
+  []
+  (when-let [keyset-value @keyset]
+    (keys keyset-value)))
 
 (defn- group-by-kid [certs]
  (->> (for [{:keys [kid] :as cert} certs]
         [kid (bk/jwk->public-key cert)])
-      (into {})
-      (reset! keyset)))
+      (into {})))
 
 (defn- parse-jwk-response [{:keys [body] :as response}]
   (try
-    (-> (j/read-value body t/mapper)
-        :keys
-        group-by-kid)
+    (->> (json/read-value body t/mapper)
+         :keys
+         group-by-kid
+         (reset! keyset))
     (catch Exception ex
       (-> "unable to retrieve key from jwk response"
           (ex-info {:response response} ex)
           throw))))
 
-(defn jwks->keys [url]
+(defn- jwks->keys [url]
   (let [response @(http/get url)]
     (if-not (response :error)
       (parse-jwk-response response)
       (-> (str "unable to call jwk url: " url)
           (ex-info response)
           throw))))
 
-(defn find-token-key [jwks token]
+(defn refresh-kids
+  "Refresh new set of jwks-keyset from auth-provider"
+  [uri callback]
+  (let [refreshed-keyset (jwks->keys uri)]
+    (prn "[funcade]: refreshed keyset")
+    (callback (keys refreshed-keyset))
+    (keys refreshed-keyset)))
+
+(defn jwks->keys-fn
```

Comment:
it should not couple:
* getting the keyset
* setting the initial state
* sheduling the refresh

and
* returning a function it does not create (or knows nothing about)

this function should be called "`jwks->keyset`", and it should do:
* `jwks->keys`
* schedule

that's it